### ISSUE TITLE
Changing API keys does not go well

### ISSUE THREAD

**Issue by autoSteve** [BJReplay/ha-solcast-solar#183]
Impacts later versions of the code.

**(Note: A fix will be pushed in v4.2.1.)**

If one changes API key (edit: should be "changes to a new account and API key and different site IDs") then it is an almost certainty that everything Solcast will turn to custard. It turned to custard in spectacular fashion for me whilst testing this rare happening.

The logs were quite breathtaking, if I'm honest. I won't share, despite their cruel beauty.

The fix is to remove all solcast JSON cache files from the config directory, scorched-earth-reset style, and then re-load.

Please don't try an API key (edit: Solcast account) change for shits and giggles to replicate this. Unless you backed up the instant prior, have sadistic tendancies and want to watch the show...

The fix is being worked on.

(Further, if the API key (edit: account/key) is changed then re-load happens, and if the integration just happens to hit a '429 go away we're busy' then it will blindly load the previous sites as if they belonged to the new API key (edit: and account) and bad stuff will not immediately happen. Rather, this will result in tears and an exploded log some time later.)

**Comment by BJReplay** [comment]
Ok, I am intrigued. Different API key, but same resource IDs? Or both change?

**Comment by autoSteve** [comment]
If an API key changes, then the site IDs _will_ change. Yep. Deffo.

But the crux to the custard is an API change.

I am making headway. (Apologies for bumping your usage while I switch between that and the Sydney Opera House, violating Solcast's terms and conditions just a weeny little bit for the time being... Their DOS detection has slowed me down somewhat, giving me 429s for a whole half hour! Unmetered my arse...)

**Comment by BJReplay** [comment]
> Apologies for bumping your usage while I switch between that and the Sydney Opera House

No worries - it is there for the common good.

**Comment by autoSteve** [comment]
Righto.

API key change is golden. No logs hysterically spammed (awww! it was awesome!!!). There are now not enough 'O's in smooth to describe the experience.

What is not yet golden is that when the API key gets changed when one is using just one API key, as is Solcast's desire and want and T's&C's blah, blah, the API usage does not get reset to zero. It should be really, because you've got a new API key.

This is doing my head in somewhat, currently seeking a coding epiphany, but I will hurdle that... Maybe today. Maybe tomorrow.

**Comment by autoSteve** [comment]
It is officially bed time... I got a limit exceeded from Solcast for a supposedly **unmetered** site... The Sydney Opera House. Gah! Then bad things.

```
2024-10-14 23:51:04.187 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Configuration directory is /config
2024-10-14 23:51:04.205 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Sites cache exists
2024-10-14 23:51:04.205 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Connecting to https://api.solcast.com.au/rooftop_sites?format=json&api_key=******sxdNS5
2024-10-14 23:51:04.343 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] HTTP session returned status 200/Success in __sites_data()
2024-10-14 23:51:04.343 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Writing sites cache
2024-10-14 23:51:04.355 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Sites data: {'sites': [{'name': 'Sydney Opera House', 'resource_id': '69d4-4611-63a4-5baa', 'capacity': 13.2, 'capacity_dc': 12, 'longitude': **.******, 'latitude': **.******, 'azimuth': 90, 'tilt': 30, 'install_date': '2024-10-13T13:00:00.0000000Z', 'loss_factor': 0.9, 'tags': [], 'location': 'Sydney, NSW Australia', 'api_key': '******sxdNS5'}], 'page_count': 1, 'current_page': 1, 'total_records': 1}
2024-10-14 23:51:04.356 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Usage cache exists for ******sxdNS5
2024-10-14 23:51:04.365 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Usage cache for ******sxdNS5 last reset 2024-10-14 11:00:00
2024-10-14 23:51:04.366 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] API counter for ******sxdNS5 is 17/48
2024-10-14 23:51:04.366 DEBUG (MainThread) [custom_components.solcast_solar] Successful init
2024-10-14 23:51:04.366 INFO (MainThread) [custom_components.solcast_solar] Solcast integration version: 4.2.0
2024-10-14 23:51:04.384 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Using legacy hourly dampening
2024-10-14 23:51:04.432 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Data cache /config/solcast.json exists, file type is <class 'dict'>
2024-10-14 23:51:04.450 INFO (MainThread) [custom_components.solcast_solar.solcastapi] New site(s) have been added, so getting forecast data for them
2024-10-14 23:51:04.450 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Polling API for site 69d4-4611-63a4-5baa lastday 2024-10-21 numhours 169
2024-10-14 23:51:04.450 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Fetch data url: https://api.solcast.com.au/rooftop_sites/69d4-4611-63a4-5baa/estimated_actuals
2024-10-14 23:51:04.451 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Fetching forecast
2024-10-14 23:51:04.857 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Fetch successful
2024-10-14 23:51:04.858 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] API returned data, API counter incremented from 17 to 18
2024-10-14 23:51:04.858 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Writing API usage cache file: /config/solcast-usage.json
2024-10-14 23:51:04.867 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] HTTP session returned data type <class 'dict'>
2024-10-14 23:51:04.867 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] HTTP session status 200/Success
2024-10-14 23:51:04.989 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Fetch data url: https://api.solcast.com.au/rooftop_sites/69d4-4611-63a4-5baa/forecasts
2024-10-14 23:51:04.989 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Fetching forecast
2024-10-14 23:51:05.042 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Writing API usage cache file: /config/solcast-usage.json
2024-10-14 23:51:05.058 ERROR (MainThread) [custom_components.solcast_solar.solcastapi] API allowed polling limit has been exceeded, API counter set to 48/48
2024-10-14 23:51:05.059 WARNING (MainThread) [custom_components.solcast_solar.solcastapi] Site resource id a546-98c5-53c5-d101 is no longer configured, removing saved data from cached files /config/solcast.json, /config/solcast-undampened.json
2024-10-14 23:51:05.068 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Saved forecast cache /config/solcast.json
2024-10-14 23:51:05.100 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Saved forecast cache /config/solcast-undampened.json
2024-10-14 23:51:05.100 INFO (MainThread) [custom_components.solcast_solar.solcastapi] Migrating undampened history to /config/solcast-undampened.json for 69d4-4611-63a4-5baa
2024-10-14 23:51:05.131 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Saved forecast cache /config/solcast-undampened.json
2024-10-14 23:51:05.132 INFO (MainThread) [custom_components.solcast_solar.solcastapi] Dampening forecasts for today onwards for site 69d4-4611-63a4-5baa
2024-10-14 23:51:05.140 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Saved forecast cache /config/solcast.json
2024-10-14 23:51:05.146 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Data cache /config/solcast-undampened.json exists, file type is <class 'dict'>
2024-10-14 23:51:05.156 ERROR (MainThread) [custom_components.solcast_solar.solcastapi] Exception in get_forecast_update(): Traceback (most recent call last):
  File "/config/custom_components/solcast_solar/solcastapi.py", line 2501, in build_forecast_data
    await self.__check_data_records()
  File "/config/custom_components/solcast_solar/solcastapi.py", line 2538, in __check_data_records
    st_i, end_i = self.__get_forecast_list_slice(self._data_forecasts, start_utc, end_utc)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/solcast_solar/solcastapi.py", line 1671, in __get_forecast_list_slice
    _forecasts_start_idx = self.__calc_forecast_start_index(_data)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/solcast_solar/solcastapi.py", line 2530, in __calc_forecast_start_index
    return idx
           ^^^
UnboundLocalError: cannot access local variable 'idx' where it is not associated with a value
```

**Comment by autoSteve** [comment]
Turns out...

> _Please note that Unmetered Locations are not available for the Legacy products including Weather Sites, Utility Scale Sites, **Rooftop Sites** and World API._

The hobbyist rooftop sites is legacy, so I can't even rely on the opera house. No free lunch, so fixing this will be significantly slowed down to the rate of ten API calls per day.

<img width="217" alt="image" src="https://github.com/user-attachments/assets/4645ec5f-367e-4595-b58f-ab020d426acd">

**Comment by BJReplay** [comment]
+20 of mine (and if you go over, I don't care)

**Comment by autoSteve** [comment]
This issue is resolved.

The fix will be included in v4.2.1 when that gets released. Leaving this issue open until that happens.
